### PR TITLE
Moment Interaction Diagram Usability Enhancements

### LINKED ISSUE #39: Moment Interaction Diagram Usability Enhancements
- More helpful exception messages when number of ``labels``, ``control_points`` or ``n_points`` do not match
- Automatic sorting of ``control_points``, rather than raising an exception if not sorted

_Originally posted by @Agent6-6-6 in https://github.com/robbievanleeuwen/concrete-properties/discussions/32#discussioncomment-3492362_

### PR DESCRIPTION
**Pull request by robbievanleeuwen**
Fixes #39.

Introduces/redefines the following parameters to ``moment_interaction_diagram()``:
- ``limits`` - defines the limits of the moment interaction diagram (note there are some cases e.g. AS3600 where I don't want to generate max compression to max tension)
- ``n_points`` - now defines the number of points between the limits
- ``n_spacing`` - Optional axial force spacing for the moment interaction diagram

Other features:
- Sorting of ``control_points`` no longer required, ~~only requirement is that ``limits`` be in order from larger ``N`` to lower ``N`` (error message when not the case) - this is due to how ``kappa0`` is implemented which effectively implements a neutral axis depth of infinity.~~ EDIT: no longer required.

### CODE REVIEW

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6 this is still a WIP. Outstanding tasks:

- [ ] Update all examples and other methods that call ``moment_interaction_diagram()``
- [ ] Update ``max_comp`` with fix as suggested in #39
- [ ] Resolve ``n_points`` spacing issue

**Comment by robbievanleeuwen** [comment]
**``n_points`` Spacing Issue:**

> Either have a user supplied variable to break this range up evenly by a certain number of evenly spaced control points (like ``n_points = 48``)...

This creates the following issue:

image

Points are focussed towards higher axial loads as setting ``kappa0`` as a limit point equates to a neutral axis depth of ``2 * D``.

Wondering if you have any thoughts as to how this should behave? Haven't thought too much about it myself yet.

**Comment by robbievanleeuwen** [comment]
**Axial Force Spacing:**

This seems to be looking good!

image

**Comment by robbievanleeuwen** [comment]
**Axial Force Spaing:**

On the other hand, this seems to be working well!
image

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6 I've changed the default arguments to get better behaviour for default calls of ``moment_interaction_diagram()``.

image

The limits now go from decompression ``("D", 0)`` to full tension ``("d_n", 1e-6)`` with a control point added at full compression ``("kappa0", 0)`` as well as balanced and pure bending. This causes the neutral axes to be spaced between the decompression and tension point which has a better distributed default curve.

If the user wants more points towards the full compression they can easily change their limits and/or use ``n_spacing`` (of course this will take much longer to generate but is *prettier*). There are no longer any requirements to put things in a particular order. I think this is better behaviour.

image

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6, this one is ready for review. Would love to get your thoughts on whether this makes moment interaction diagrams more usable!

**Comment by Agent6-6-6** [comment]
@robbievanleeuwen one comment so far before taking it for a spin, you're referring to the limits with the decompression limit point being based on the ratio of neutral axis depth to section depth = "D". 

But I think the actual point at which the concrete section starts to not be fully in compression is actually when the neutral axis depth is `h/gamma` if h = section depth and gamma is the factor by which you're relating the neutral axis depth to compression block depth in your code nomenclature. Maybe I'm misinterpreting though.

But obviously, this would depend on the stress-strain relationship, but it seems like you're potentially missing the true decompression point as implemented if you're solving for a neutral axis depth equal to section depth as the uppermost limit? 

To demonstrate for example if gamma = 0.85, and h = 500mm, then the decompression point is really when the neutral axis depth = 500/0.85 = 588 mm. This came up in #26 when you reworked the interaction diagrams before.

If I'm correct here you'd need to somehow determine where the stress-strain profile transitions to the constant plateau (for rectangular, bilinear and parabolic stress-strain curves) and work out the upper bound limit based on that to achieve the full section in compression?

**Comment by Agent6-6-6** [comment]
One other observation @robbievanleeuwen, with the `n_spacing` parameter related to axial load step, since you don't know the total extent between limits you kind of have to run a few times to tune the value of `n_spacing` to get the desired result. Underestimate it and you might be in for a longer analysis than expected.

I kind of wonder if the `n_spacing` parameter should really be defined as a number of equally spaced axial loads to keep (irrespective of the section and range of axial load between the limits), which hopefully keeps the number of iterations similar irrespective of the section being analysed. As opposed to a hard-coded axial load value like 200kN which will give a different accuracy of the interaction curve for sections with 10000kN axial range vs 500kN axial range along with quite significant differences in run time.

Or alternatively, consider adding this as a third option for controlling the resulting output for the curve on top of n_points and n_spacing?

File: docs/source/notebooks/moment_interaction.ipynb [hunk 1]
```diff
@@ -282,24 +282,55 @@
...
+    "The following parameters can be used to modify the moment interaction diagram:\n",
+    "\n",
+    "1. ``limits`` - List of control points that define the start and end of the interaction diagram. List length must equal two. The default limits range from decompression ``(\"D\", 1.0)`` to zero curvature tension ``(\"d_n\", 1e-6)``.\n",
+    "\n",
+    "2. ``control_points`` - adding control points to the moment interaction diagram allows the user to ensure certain points are captured on the moment interaction diagram.  The default argument for ``control_points`` is ``[(\"kappa0\", 0.0), (\"fy\", 1.0), (\"N\", 0.0)]`` which produces control points at the pure compression point, the balanced point and the pure bending point. \n",
+    "    \n",
+    "\n",
+    "3. ``labels`` - A list of labels to apply to the ``limits`` and ``control_points`` for plotting purposes. The first two values in ``labels`` apply labels to the ``limits``, the remaining values apply labels to the ``control_points``.\n",
+    "\n",
+    "4. ``n_points`` - The number of points to compute including and between the ``limits`` of the moment interaction diagram.\n",
```
Comment (Agent6-6-6) [inline docs/source/notebooks/moment_interaction.ipynb:306]: add here that `n_points` is number of points based on evenly spaced neutral axis depths?

File: docs/source/notebooks/moment_interaction.ipynb [hunk 2]
```diff
@@ -282,24 +282,55 @@
...
-    "The below example will use all these features, creating additional control points at zero strain, 50% strain and a neutral axis depth of 200 mm. Note, care must be taken to ensure that the ``control_points`` are ordered from largest neutral axis depth to smallest neutral axis depth otherwise a ``ValueError`` will be raised. Labels are applied and the number of points between each control point is modified. A maximum compressive force of 6800 kN is provided."
+    "The below example will use these features, creating additional control points at:\n",
+    "\n",
+    "- Decompression of the concrete ``(\"D\", 1.0)`` - label ``C``\n",
+    "- Decompression of the steel ``(\"fy\", 0.0)`` - label ``D``\n",
+    "- 50% yield strain in the steel ``(\"fy\", 0.5)`` - label ``E``\n",
+    "- 100% yield strain in the steel ``(\"fy\", 1.0)`` - label ``F``\n",
+    "- Neutral axis depth of 200 mm ``(\"d_n\", 200.0)`` - label ``G``\n",
+    "\n",
+    "A maximum compressive force of 6800 kN is provided. Label ``A`` is placed at zero axial force and label ``B`` is placed at the intersection with the moment interaction diagram.\n",
```
Comment (Agent6-6-6) [inline docs/source/notebooks/moment_interaction.ipynb:326]: Label point A is at zero moment, not zero axial load I think.

File: docs/source/user_guide/analysis.rst [hunk 3]
```diff
@@ -122,14 +122,14 @@ method. By default the moment interaction diagram is generated for bending about
   :noindex:
 
 The moment interaction diagram is generated by shifting the neutral axis throughout the
-cross-section from pure bending to the decompression point using ``n_points``. A
-straight line is generated between the decompression point and the squash load, and the
-pure bending point and the tensile load.
+cross-section between the ``limits`` using either ``n_points`` or ``n_spacing``.
```
Comment (Agent6-6-6) [inline docs/source/user_guide/analysis.rst:125]: wonder if this should be expended on a little to illustrate the difference between the two options

```suggestion
cross-section between the ``limits`` using either ``n_points`` number of equally spaced neutral axis depths or by determining neutral axis depths relating to ``n_spacing`` number of equally spaced axial load analysis points.
```

File: docs/source/notebooks/moment_interaction.ipynb [hunk 4]
```diff
@@ -282,24 +282,55 @@
...
-    "    e. ``kappa_0`` - zero curvature compression, this must be at start of the list as it results in the largest compression force. Note that the second value in the tuple is not used and can be anything.<br>\n",
+    "The limits of the moment interaction diagram and any additional analysis points can be specified by defining control points. Control points are defined as a list of (``str``, ``float``) tuples. The following types of control points are used in ``concreteproperties``:\n",
     "\n",
-    "2. ``labels`` - A list of labels to apply to the ``control_points`` for plotting purposes.\n",
-    "3. ``n_points`` - The number of neutral axis depths to compute between each control point.\n",
-    "3. ``max_comp`` - Limits the maximum compressive force in the moment interaction diagram by cutting off the top of the diagram."
+    "1. ``D`` - ratio of neutral axis depth to section depth, e.g. the (``\"D\"``, ``0.5``) control point will ensure a neutral axis at mid-depth is analysed.\n",
+    "\n",
+    "2. ``d_n`` - neutral axis depth, e.g. the (``\"d_n\"``, ``200``) control point will ensure a netural axis at 200 is analysed.\n",
```
Comment (Agent6-6-6) [inline docs/source/notebooks/moment_interaction.ipynb:289]:
```suggestion
    "2. ``d_n`` - neutral axis depth, e.g. the (``\"d_n\"``, ``200``) control point will ensure a neutral axis at 200 is analysed.\n",
```

File: concreteproperties/concrete_section.py [hunk 5]
```diff
@@ -1036,186 +1036,215 @@ def calculate_ultimate_section_actions(
...
+        # get neutral axis depths for additional control points
+        add_cp_dn = []
+
+        for cp in control_points:
+            add_cp_dn.append(self.decode_d_n(theta=theta, cp=cp, d_t=d_t))
 
         # validate labels length
-        if len(labels) != len(control_points):
-            raise ValueError("Length of labels must equal length of control_points.")
+        if labels and len(labels) != 1 and len(labels) != 2 + len(control_points):
```
Comment (Agent6-6-6) [inline concreteproperties/concrete_section.py:1128]: should this be with an or as implied in the valueerror text?:-
```suggestion
        if labels and (len(labels) != 1 or len(labels) != 2 + len(control_points)):
```

File: concreteproperties/concrete_section.py [hunk 6]
```diff
@@ -1036,186 +1036,215 @@ def calculate_ultimate_section_actions(
...
-            label_list.extend([None] * (n_pt - 2))
+            # add end axial force
+            analysis_list.append(end_res.n)
+        else:
+            # check for infinity in limits - this will not work with linspace
+            # for sake of distributing neutral axes let kappa0 ~= 2 * D
+            if limits_dn[0] == inf:
+                start = 2 * d_t
+            else:
+                start = limits_dn[0]
```
Comment (Agent6-6-6) [inline concreteproperties/concrete_section.py:1172]: should there be a check if the `limits` are supplied the wrong way around prior to calculating `limits_dn` above and swap or just use `max(limits)` and `min(limits)` so the user entered order is not important? (rather than implying `limits[0]` and `limits[1]` are max and min)

File: concreteproperties/concrete_section.py [hunk 7]
```diff
@@ -1240,6 +1269,12 @@ def micurve(progress=None):
 
         # cut diagram at max_comp
         if max_comp:
+            # check input - if value greater than maximum compression
+            if max_comp > mi_results.results[0].n:
+                msg = f"max_comp={max_comp} is greater than the maximum axial load "
```
Comment (Agent6-6-6) [inline concreteproperties/concrete_section.py:1274]: think this should be axial capacity, not load?

```suggestion
                msg = f"max_comp={max_comp} is greater than the maximum axial capacity "
```

File: concreteproperties/concrete_section.py [hunk 8]
```diff
@@ -1258,6 +1293,7 @@ def micurve(progress=None):
                 # determine which index is the first to keep
                 if idx_to_keep == 0 and mi_res.n < max_comp:
                     idx_to_keep = idx
+                    break
 
             # create interpolation function and determine moment which corresponds to
```
Comment (Agent6-6-6) [inline concreteproperties/concrete_section.py:1298]: Rather than interpolating here, why not just calculate neutral axis depth `d_n` at `max_comp` and then work out actual moment capacity at this control point. 

I wonder if `max_comp` used, maybe easier to add another artificial control point at this axial load so you calculate the moment capacity value in above code iterations when going through all the points, then you're just cutting off the unused values above max_comp here?

This would be more accurate.

File: concreteproperties/design_codes/as3600.py [hunk 9]
```diff
@@ -437,37 +436,48 @@ def non_linear_phi(phi_guess):
...
-            purposes, length must be the same as the length of ``control_points``. If a
-            single value is provided, will apply this label to all control points.
-        :param n_points: Number of neutral axis depths to compute between each control
-            point. Length must be one less than the length of ``control_points``. If an
-            integer is provided this will be used between all control points.
+        :param limits: List of control points that define the start and end of the
+            interaction diagram. List length must equal two. The default limits range
+            from decompression to zero curvature tension.
+        :param control_points: List of additional control points to add to the moment
+            interatction diagram. The default control points include the pure
```
Comment (Agent6-6-6) [inline concreteproperties/design_codes/as3600.py:464]:
```suggestion
            interaction diagram. The default control points include the pure
```

File: concreteproperties/design_codes/as3600.py [hunk 10]
```diff
@@ -437,37 +436,48 @@ def non_linear_phi(phi_guess):
...
+            for plotting purposes. The first two values in ``labels`` apply labels to
+            the ``limits``, the remaining values apply labels to the ``control_points``.
+            If a single value is provided, this value will be applied to both ``limits``
+            and all ``control_points``. The length of ``labels`` must equal ``1`` or
+            ``2 + len(control_points)``.
+        :param n_points: Number of points to compute including and between the
+            ``limits`` of the moment interaction diagram. Generates equally spaced
+            neutral axes between the ``limits``.
+        :param n_spacing: If provided, overrides ``n_points`` and spaces the points on
+            the moment interaction diagram by an axial force ``n_spacing``. Note that
```
Comment (Agent6-6-6) [inline concreteproperties/design_codes/as3600.py:478]:
```suggestion
            the moment interaction diagram by a constant axial force ``n_spacing``. Note that
```

File: docs/source/notebooks/moment_interaction.ipynb [hunk 11]
```diff
@@ -282,24 +282,55 @@
...
   },
   {
    "cell_type": "markdown",
    "id": "303c4420",
    "metadata": {},
    "source": [
-    "The below example will use all these features, creating additional control points at zero strain, 50% strain and a neutral axis depth of 200 mm. Note, care must be taken to ensure that the ``control_points`` are ordered from largest neutral axis depth to smallest neutral axis depth otherwise a ``ValueError`` will be raised. Labels are applied and the number of points between each control point is modified. A maximum compressive force of 6800 kN is provided."
+    "The below example will use these features, creating additional control points at:\n",
+    "\n",
+    "- Decompression of the concrete ``(\"D\", 1.0)`` - label ``C``\n",
```
Comment (Agent6-6-6) [inline docs/source/notebooks/moment_interaction.ipynb:320]:
```suggestion
    "- Decompression point for the concrete ``(\"D\", 1.0)`` - label ``C``\n",
```

File: docs/source/notebooks/moment_interaction.ipynb [hunk 12]
```diff
@@ -282,24 +282,55 @@
...
+    "- Decompression of the steel ``(\"fy\", 0.0)`` - label ``D``\n",
```
Comment (Agent6-6-6) [inline docs/source/notebooks/moment_interaction.ipynb:321]:
```suggestion
    "- Decompression point for the steel ``(\"fy\", 0.0)`` - label ``D``\n",
```

**Review by Agent6-6-6 — CHANGES_REQUESTED** [review]
@robbievanleeuwen added a few suggestions/corrections regarding docs and some other queries/suggestions. Let me know if anything isn't clear!

File: concreteproperties/concrete_section.py (hunk 5 above)
Comment (Agent6-6-6) [reply concreteproperties/concrete_section.py:1128]: Actually, I think your logic is correct here. Was reading it as equal to!

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6 

> @robbievanleeuwen one comment so far before taking it for a spin, you're referring to the limits with the decompression limit point being based on the ratio of neutral axis depth to section depth = "D".

I think my approach here will be to change the docs such that it is clear that it is the strain profile that is in decompression, not necessarily the stress (as you say, depends on your stress-strain profile). Given that there are many possible stress-strain profiles (not just the rectangular stress block), compounded with the possibly of sections with differing grades of concrete, I'm not going to try to work out when the stress "decompresses", but rather work with strain. 

Design codes and/or users can then modify this value to their needs if they require.

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6 

> One other observation @robbievanleeuwen, with the n_spacing parameter related to axial load step, since you don't know the total extent between limits you kind of have to run a few times to tune the value of n_spacing to get the desired result.

This is a fair comment, I will add an additional parameter to allow for ``n`` evenly spaced axial loads.

File: docs/source/user_guide/analysis.rst (hunk 3 above)
Comment (robbievanleeuwen) [reply docs/source/user_guide/analysis.rst:125]: Have moved this paragraph above the ``automethod``. Won't elaborate on ``n_points`` and ``n_spacing`` here as its covered in the docstring.

File: docs/source/notebooks/moment_interaction.ipynb (hunk 2 above)
Comment (robbievanleeuwen) [reply docs/source/notebooks/moment_interaction.ipynb:326]: Yes good pickup

File: docs/source/notebooks/moment_interaction.ipynb (hunk 1 above)
Comment (robbievanleeuwen) [reply docs/source/notebooks/moment_interaction.ipynb:306]: Thanks, changed.

File: concreteproperties/concrete_section.py (hunk 6 above)
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:1170]: There is another ``if`` statement below this that handles this. I think it currently works letting you enter limits in either order. Let me know if you have an example that breaks this.

File: concreteproperties/concrete_section.py (hunk 7 above)
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:1272]: Yes good point!

File: concreteproperties/design_codes/as3600.py (hunk 9 above)
Comment (robbievanleeuwen) [reply concreteproperties/design_codes/as3600.py:464]: Thanks!

File: concreteproperties/concrete_section.py (hunk 8 above)
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:1296]: This is a far more sensible approach, thanks. I was approaching this as a (bad) programmer not an engineer 🤣 

Before I implement this, just wanting to get clarification on something - should the first point in the diagram be at zero moment or have the same moment as the pretrimmed diagram, i.e. point 1 or 2 in the below? Current implementation is point 2.

Screenshot 2022-11-07 130727

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6 another off-topic one if you don't mind 😄 Wondering what hardware you use to run ``pfSense`` (had a quick little read of [this](https://engineervsheep.com/2019/the-ruskies-are-coming/)!)

**Comment by Agent6-6-6** [comment]
> @Agent6-6-6 another off-topic one if you don't mind 😄 Wondering what hardware you use to run `pfSense` (had a quick little read of [this](https://engineervsheep.com/2019/the-ruskies-are-coming/)!)

@robbievanleeuwen - I run it on a little fanless Qotom PC similar to [this ](https://www.aliexpress.com/item/32769767156.html?spm=a2g0o.detail.100009.5.72d91011NfLdZj&gps-id=pcDetailLeftTopSell&scm=1007.13482.271138.0&scm_id=1007.13482.271138.0&scm-url=1007.13482.271138.0&pvid=6ac49b28-fecf-4738-aeb2-467500c7ab62&_t=gps-id:pcDetailLeftTopSell,scm-url:1007.13482.271138.0,pvid:6ac49b28-fecf-4738-aeb2-467500c7ab62,tpp_buckets:668%232846%238115%232000&pdp_ext_f=%7B%22sku_id%22%3A%2212000025891370461%22%2C%22sceneId%22%3A%223482%22%7D&pdp_npi=2%40dis%21USD%21164.0%21155.8%21%21%21%21%21%402101d1ac16678015129334132eef24%2112000025891370461%21rec)but from back in 2017... just have it screwed to the underside of a shelf where my fibre connection terminates inside the house. Took a few weeks to arrive from China form memory

image
4Gb of memory, 32Gb mSSD.

It does gets a bit toasty in Summer on hot days. But it is more than capable of maxing out my 300/100Mbit fibre connection if required. Took a while to get my head around how all the rules work for the firewall and setting up other features. But once over that hump, I'd never go back to using a consumer router. 

You can get it all set up for a lot less cost than a half-decent high-end router that would offer you a lot less control over tinkering with it! Plus pfSense is regularly updated form a feature and security standpoint since it's mostly open source.

File: concreteproperties/concrete_section.py (hunk 8 above)
Comment (Agent6-6-6) [reply concreteproperties/concrete_section.py:1288]: @robbievanleeuwen, I'd say zero moment (point 2), simply because if you plotted the negative side of the diagram it all meets up as a horizontal line anyway to complete the full interaction surface slice. So the actual value isn't that important as at that axial load you know you can have a moment between whatever the capacity is at `-M_max_comp` or `+M_max_comp`.

I'll have a look at what SPColumn does though to see if there is another precedent.

I was thinking that this additional point just gets appended to the evenly spaced points, rather than a control point so as not to mess up the labels?

I wonder if `max_comp` is specified if that should then substitute in for the upper limit so as not to waste computation time working out points above this axial load (don't recall maybe you already did this, been a week since I did the review!?). I'm thinking here for wall design in NZ at least, the max compression load is like `0.3f'cAg` which is usually way below the balance point, so there is no point working out data points above this only to then truncate the results and throw a large percentage of them away

**Comment by Agent6-6-6** [comment]
> 



> I think my approach here will be to change the docs such that it is clear that it is the strain profile that is in decompression, not necessarily the stress (as you say, depends on your stress-strain profile). Given that there are many possible stress-strain profiles (not just the rectangular stress block), compounded with the possibility of sections with differing grades of concrete, I'm not going to try to work out when the stress "decompresses", but rather work with strain.
> 
> Design codes and/or users can then modify this value to their needs if they require.
> 
> EDIT: Because the rectangular stress block attempts to simplify the "real" concrete stress profile I feel like decompression should still be based on the strain profile and not the stress profile. This way if you are comparing decompression of a rectangular stress block with a parabolic stress block you _should_ get similar results. Pushing down the rectangular stress block by `D/gamma` wouldn't make sense in this case as your strain profile no longer corresponds to decompression and comparison with a parabolic profile at decompression would be way off.

I've always thought the decompression point was where the concrete in the section started to not be fully in compression at the design compression stress. Independent of the stress-strain profile this is still the same physical phenomenon in terms of cross-section stresses. i.e. for a rectangular stress block when the neutral axis is `D/gamma`, for a parabolic stress block when `D/(ratio of distance on strain diagram where it transitions to the horizontal to the maximum strain)`, similar for a bilinear approach. For all of the ultimate stress-strain profiles, I think you already have these values probably as they are used to generate the stress-strain profile in the first place. 

Can follow that it would be a bit of a dog with multiple different concrete sections with different strengths and hence different parameters like gamma (still have not wrapped my head around how this is even done for the ULS analysis yet!)

Anyway, I'll see what you say in the docs as probably not fully understanding exactly what has been done, just seemed like you're missing out on a bit of the curve between the max point and this first calculated data point where you take the neutral axis depth at `D`, but granted nothing interesting usually happens at the top of the curve anyway and in my use case that part of the curve usually seems to be truncated by `max_comp`!!! 

I'm not so sure we are talking about the same thing, as not really envisaging how the strain comes into it when it is a stress based thing, i.e. the point at which the stress starts to reduce below a constant concrete stress (for the concrete only) across the entire section? Past this point obviously only the bar forces can vary to take more load in compression.

File: concreteproperties/concrete_section.py (hunk 8 above)
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:1288]: Ok thanks for that, some good points, will implement!

**Comment by robbievanleeuwen** [comment]
Thanks for the heads up on the aliexpress special, everything else I was looking at seemed very expensive! Yep the open source nature and features pfSense offer are very appealing!

Yeah maybe we are thinking of two different things, decompression for me is when one extreme fibre of the concrete has a stress of zero (and thus a corresponding strain of zero for most stress-strain profiles).

The below picture is by no means definitive, just one of the first results on google, but kind of explains what I probably haven't put into words very eloquently. Zero concrete strain at the bottom, and for a rectangular stress block some portion of the cross-section with zero stress.

Setting ``("D", 1.0)`` as a control point puts the neutral axis at a depth ``D`` and thus gives you this "decompression point". Agreed you lose a bit of fidelity between this point and pure compression, but the user can easily change the limits if they desire. See the very first image in this thread if you change the top limit to ``("D", 2.0)`` and use ``n_points``, not very nicely spaced, hence why I went with ``("D", 1.0)``.

image

**Comment by Agent6-6-6** [comment]
OK got it now. Zero strain at extreme fibre = decompression point. 

Yeah plenty of options to tweak if people want to refine the curve above the decompression point if there is any markedly different behaviour going on, so all good!

Its kind of got me thinking if there should also be a control point behaviour based on defining strain at the extreme tension (concrete) fibre. You have one for stell with the ʼf_yʼ for example. Though not sure how much use it is but I guess it is just a different way of specifying the  ʼDʼ ratio that might seem a bit more logical with respect to the analysis actually being strain based. I mean would basing the step on equally spaced strains vs equally spaced neutral axis depths just be a same same but different thing?

File: concreteproperties/concrete_section.py (hunk 8 above)
Comment (Agent6-6-6) [reply concreteproperties/concrete_section.py:1288]: Hi @robbievanleeuwen 

> I'll have a look at what SPColumn does though to see if there is another precedent.

I had a look at this but forgot to post back, what SPColumn does if you look at the positive or negative branches of the diagram is just cut it off at zero moment at the axes, but has the part of the negative curve generated at max compression or max tension, for example +ve and -ve views below:-
image
image

So you get the full half of the diagram, not a skewed result for a non-symmetric section or reinforcement with different moment capacities top and bottom like concreteproperties returns. That could lead to not quite knowing for example if in your last diagram you're at an N*=-350kN and M*=25kNm that you're above or below the curve even though you're in the positive half of the interaction curve. 

As a thought I wonder how hard would it be to generate the remaining missing points back to zero moment in the tension region in your last example and to have the option of truncating the curve at zero moment to implement something similar to SPColumn for a consistent approach to a positive and negative portion of the curve. This would also make it easier to test if inside the curve or not if you added a closing line at zero moment and then used some shapely stuff to check if a point lies within a polygon or similar.

Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py]: Thanks for exploring this. I'm going to turn this into a new issue as it raises a few items and it would be nice to get this PR merged for now.
